core/store/src/trie/update.rs

```diff
@@ -33,6 +37,33 @@ impl TrieUpdate {
             self.trie.get(&self.root, key).map(|x| x.map(DBValue::from_vec))
         }
     }
+
+    /// Get values in trie update for a set of keys.
+    /// Returns: a hash map of prefix -> <key, value> changes in the trie update.
+    // This function will commit changes. Need to be used with caution
+    pub fn get_prefix_changes(
+        &mut self,
+        prefixes: &HashSet<Vec<u8>>,
+    ) -> Result<PrefixKeyValueChanges, StorageError> {
+        if !self.prospective.is_empty() {
+            panic!("Uncommitted changes exist");
+        }
+        let mut res = HashMap::new();
+        for prefix in prefixes {
+            let mut prefix_key_value_change = HashMap::new();
+            for (key, value) in self.committed.range(prefix.to_vec()..) {
```

Comment:
A minor optimization: We could create an `end` by flipping the right-most zero in `prefix` to 1 and flipping all 1ones after that to 0 (If there are no zeros then we iterate to the end).
For instance, if prefix=start=[0,0,1,0,1,1,1] then end=[0,0,1,1,0,0,0].
This will also remove the prefix check that we perform on each iteration.

Then we can do this:
```rust
prefixes.iter().map(|prefix| {
 (prefix.cloned(), self.committed.range(get_bounds(prefix)).cloned().collect())
}).collect();
```

Reply:
There are actually more optimizations we can do. Right now the main performance issue is that there might be values cloned multiple times if one prefix is a prefix of another prefix. We can do optimizations later when we actually get the subscriptions to work.